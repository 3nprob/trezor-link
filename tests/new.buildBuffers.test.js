const parseConfigure = require('../src/lowlevel/protobuf/parse_protocol-new').parseConfigure
const buildBuffers = require('../src/lowlevel/send-new').buildBuffers;
const receiveAndParse = require('../src/lowlevel/receive-new').receiveAndParse;

const messages = require('../messages.json');
const fixtures = require('./__fixtures__/messages');

const { ByteBuffer } = require("protobufjs-old-fixed-webpack");

// they also have in common the fact that we encode them, but never decode them. this makes
// me think that it a bug indeed.
const failingOnDecode = [
    'GetAddress',
    'TxAck',
    'EosTxActionAck',
    'MoneroTransactionInitRequest',
    'MoneroTransactionInitAck',
    'MoneroTransactionSetInputRequest',
    'MoneroTransactionInputViniRequest',
    'MoneroTransactionAllInputsSetAck',
    'MoneroTransactionSetOutputRequest',
    'MoneroTransactionSetOutputAck',
    'MoneroTransactionAllOutSetRequest',
    'MoneroTransactionSignInputRequest',
    'MoneroKeyImageSyncStepRequest',
    'DebugMoneroDiagRequest',
    'DebugMoneroDiagAck',
    'TezosSignTx',
    'GetOwnershipProof',
    'GetOwnershipId',
];

let parsedMessages = null
const getParsedMessages = async () => {
    if (parsedMessages) return parsedMessages;
    parsedMessages = await parseConfigure(messages);
    return parsedMessages;
}


describe('buildBuffers', () => {
    fixtures
        .filter(f => f.name === 'Features')
        .forEach(f => {

            test(`message ${f.name}`, async () => {
                const parsedMessages = await getParsedMessages();

                expect(() => {
                    buildBuffers(parsedMessages, f.name, f.params)
                }).not.toThrow();
                const result = buildBuffers(parsedMessages, f.name, f.params)
                result.forEach(r => {
                    expect(r.byteLength).toEqual(63);
                    expect(Array.from(new Uint8Array(r))).toMatchSnapshot();
                })
                console.log('result', result);
                if (!failingOnDecode.includes(f.name)) {
                    const decoded = await receiveAndParse(parsedMessages, () => {
                        return Promise.resolve(ByteBuffer.concat(result));
                    })
                    // then decode message and check, whether decoded message matches original json
                    expect(decoded.type).toEqual(f.name);
                    expect(decoded.message).toEqual(f.params);
                }
            })
        })
})

