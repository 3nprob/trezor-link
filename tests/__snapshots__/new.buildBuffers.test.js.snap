// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildBuffers message Features 1`] = `
Array [
  35,
  35,
  0,
  17,
  0,
  0,
  1,
  166,
  10,
  11,
  115,
  111,
  109,
  101,
  32,
  115,
  116,
  114,
  105,
  110,
  103,
  16,
  32,
  24,
  32,
  32,
  32,
  40,
  1,
  50,
  11,
  115,
  111,
  109,
  101,
  32,
  115,
  116,
  114,
  105,
  110,
  103,
  56,
  1,
  64,
  1,
  74,
  11,
  115,
  111,
  109,
  101,
  32,
  115,
  116,
  114,
  105,
  110,
  103,
  82,
  11,
  115,
  111,
]
`;

exports[`buildBuffers message Features 2`] = `
Array [
  109,
  101,
  32,
  115,
  116,
  114,
  105,
  110,
  103,
  96,
  1,
  106,
  64,
  133,
  31,
  201,
  84,
  35,
  66,
  50,
  26,
  246,
  62,
  203,
  186,
  125,
  62,
  206,
  84,
  95,
  42,
  66,
  186,
  208,
  27,
  163,
  44,
  255,
  85,
  53,
  177,
  142,
  84,
  182,
  211,
  16,
  110,
  16,
  182,
  164,
  82,
  89,
  147,
  209,
  133,
  161,
  68,
  61,
  154,
  18,
  81,
  134,
  150,
]
`;

exports[`buildBuffers message Features 3`] = `
Array [
  14,
  2,
  142,
  171,
  253,
  216,
  215,
  108,
  247,
  10,
  58,
  126,
  49,
  0,
  114,
  64,
  133,
  31,
  201,
  84,
  35,
  66,
  50,
  26,
  246,
  62,
  203,
  186,
  125,
  62,
  206,
  84,
  95,
  42,
  66,
  186,
  208,
  27,
  163,
  44,
  255,
  85,
  53,
  177,
  142,
  84,
  182,
  211,
  16,
  110,
  16,
  182,
  164,
  82,
  89,
  147,
  209,
  133,
  161,
  68,
  61,
  154,
  18,
]
`;

exports[`buildBuffers message Features 4`] = `
Array [
  81,
  134,
  150,
  14,
  2,
  142,
  171,
  253,
  216,
  215,
  108,
  247,
  10,
  58,
  126,
  49,
  0,
  120,
  1,
  128,
  1,
  1,
  144,
  1,
  1,
  152,
  1,
  1,
  160,
  1,
  32,
  170,
  1,
  11,
  115,
  111,
  109,
  101,
  32,
  115,
  116,
  114,
  105,
  110,
  103,
  176,
  1,
  32,
  184,
  1,
  32,
  192,
  1,
  32,
  202,
  1,
  11,
  115,
  111,
  109,
  101,
  32,
  115,
]
`;

exports[`buildBuffers message Features 5`] = `
Array [
  116,
  114,
  105,
  110,
  103,
  210,
  1,
  64,
  133,
  31,
  201,
  84,
  35,
  66,
  50,
  26,
  246,
  62,
  203,
  186,
  125,
  62,
  206,
  84,
  95,
  42,
  66,
  186,
  208,
  27,
  163,
  44,
  255,
  85,
  53,
  177,
  142,
  84,
  182,
  211,
  16,
  110,
  16,
  182,
  164,
  82,
  89,
  147,
  209,
  133,
  161,
  68,
  61,
  154,
  18,
  81,
  134,
  150,
  14,
  2,
  142,
  171,
  253,
]
`;

exports[`buildBuffers message Features 6`] = `
Array [
  216,
  215,
  108,
  247,
  10,
  58,
  126,
  49,
  0,
  216,
  1,
  1,
  224,
  1,
  1,
  232,
  1,
  1,
  240,
  1,
  1,
  248,
  1,
  0,
  128,
  2,
  1,
  136,
  2,
  1,
  144,
  2,
  1,
  154,
  2,
  64,
  133,
  31,
  201,
  84,
  35,
  66,
  50,
  26,
  246,
  62,
  203,
  186,
  125,
  62,
  206,
  84,
  95,
  42,
  66,
  186,
  208,
  27,
  163,
  44,
  255,
  85,
  53,
]
`;

exports[`buildBuffers message Features 7`] = `
Array [
  177,
  142,
  84,
  182,
  211,
  16,
  110,
  16,
  182,
  164,
  82,
  89,
  147,
  209,
  133,
  161,
  68,
  61,
  154,
  18,
  81,
  134,
  150,
  14,
  2,
  142,
  171,
  253,
  216,
  215,
  108,
  247,
  10,
  58,
  126,
  49,
  0,
  160,
  2,
  1,
  168,
  2,
  0,
  176,
  2,
  32,
  184,
  2,
  32,
  192,
  2,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;
